apps/files_sharing/lib/SharesReminderJob.php

```diff
@@ -0,0 +1,261 @@
+					$qb->expr()->lt('expiration', $qb->createNamedParameter($maxDate->format('Y-m-d H:i:s'))),
+					$qb->expr()->eq('reminder_sent', $qb->createNamedParameter(
+						false, IQueryBuilder::PARAM_BOOL
+					))
+				)
+			);
+
+		$sharesResult = $qb->executeQuery();
+		$shares = [];
+		while ($share = $sharesResult->fetch()) {
+			if ((int)$share['share_type'] === IShare::TYPE_EMAIL) {
+				$id = "ocMailShare:$share[id]";
+			} else {
+				$id = "ocinternal:$share[id]";
+			}
+
+			try {
+				$shares[] = $this->shareManager->getShareById($id);
+			} catch (ShareNotFound) {
+				$this->logger->error("Share with ID $id not found.");
+			}
+		}
+		$sharesResult->closeCursor();
+		return $shares;
+	}
+
+	/**
+	 * Checks if the user should be reminded about this share.
+	 * If so, it will retrieve and return all the necessary data for this.
+	 * It also updates the reminder sent flag for the affected shares (to avoid multiple reminders).
+	 *
+	 * @param array<IShare> $shares Shares that were obtained with {@link getShares}
+	 * @return array<array> A tuple consisting of two dictionaries: folders and languages by email
+	 * @throws Exception if the reminder sent flag could not be saved
+	 */
+	private function prepareReminders(array $shares): array {
+		// This dictionary stores email addresses as keys and folder lists as values.
+		// It is used to ensure that each user receives no more than one email notification.
+		// The email will include the names and links of the folders that the user should be reminded of.
+		$foldersByEmail = [];
+		// Similar to the previous one, this variable stores the language for each email (if provided)
+		$langByEmail = [];
+
+		/** @var IShare $share */
+		foreach ($shares as $share) {
+			if (!$this->shouldRemindOfThisShare($share)) {
+				continue;
+			}
+
+			$sharedWith = $share->getSharedWith();
+			if ($share->getShareType() == IShare::TYPE_USER) {
+				$user = $this->userManager->get($sharedWith);
+				$mailTo = $user->getEMailAddress();
+				$lang = $this->l10nFactory->getUserLanguage($user);
+				$link = $this->urlGenerator->linkToRouteAbsolute('files.view.index', [
+					'dir' => $share->getTarget()
+				]);
+			} else {
+				$mailTo = $sharedWith;
+				$lang = '';
+				$link = $this->urlGenerator->linkToRouteAbsolute('files_sharing.sharecontroller.showShare', [
+					'token' => $share->getToken()
+				]);
+			}
+			if (empty($mailTo)) {
+				continue;
+			}
+
+			if (!empty($lang)) {
+				$langByEmail[$mailTo] ??= $lang;
+			}
+			if (!isset($foldersByEmail[$mailTo])) {
+				$foldersByEmail[$mailTo] = [];
+			}
+			$foldersByEmail[$mailTo][] = ['name' => $share->getNode()->getName(), 'link' => $link];
+
+			$share->setReminderSent(true);
+			$qb = $this->db->getQueryBuilder();
+			$qb->update('share')
+				->where($qb->expr()->eq('id', $qb->createNamedParameter($share->getId())))
+				->set('reminder_sent', $qb->createNamedParameter($share->getReminderSent()))
+				->execute();
```

Comment:
```suggestion
				->executeStatement();
```
But again, I’d like to see that moved somewhere else.
The updateShare method seems a bit much for such a small flag change though, so I don’t know :thinking:

Reply:
I think updateShare works well here, the overhead (if any) is quite insignificant.